src/viper/trans.ml

```diff
@@ -323,40 +324,39 @@ and stmt ctxt (s : M.exp) : seqn =
-           let is_one = !!(EqCmpE (ghost_fld (), one)) in
-           !!(AndE (x, !!(AndE (between, !!(Implies (is_one, cond.it (exp ctxt)))))))
+           let ghost_fld () = !?(FldAcc (self ctxt cond.at, !?id)) in
+           let between = !?(AndE (!?(LeCmpE (zero, ghost_fld ())), !?(LeCmpE (ghost_fld (), one)))) in
+           let is_one = !?(EqCmpE (ghost_fld (), one)) in
+           !?(AndE (x, !?(AndE (between, !?(Implies (is_one, cond.it (exp ctxt)))))))
        | _ -> unsupported e.at (Arrange.exp e) in
      ctxt.ghost_conc := mk_c :: !(ctxt.ghost_conc);
-     !!  ([],
-          [ !!(FieldAssignS(
-              (self ctxt Source.no_region, !!id),
-              (!!(AddE(!!(FldAcc (self ctxt (s.at), !!id)),
-                       intLitE Source.no_region 1)))));
-            !!(ExhaleS (!!(AndE(!!(MacroCall("$Perm", self ctxt at)),
-                                !!(MacroCall("$Inv", self ctxt at))))));
-            !!(SeqnS (
-                let (!!) p = !!! at p in
-                !!([],
-                   [
-                     !!(InhaleS (!!(AndE(!!(MacroCall("$Perm", self ctxt at)),
-                                    !!(AndE(!!(MacroCall("$Inv", self ctxt at)),
-                                            !!(GtCmpE(!!(FldAcc (self ctxt at, !!id)),
-                                                 intLitE Source.no_region 0))))))));
-                     !!(FieldAssignS(
-                            (self ctxt at, !!id),
-                            (!!(SubE(!!(FldAcc (self ctxt at, !!id)),
-                                     intLitE at 1)))));
-                     !!! (e.at) (SeqnS stmts);
-                     !!(ExhaleS (!!(AndE(!!(MacroCall("$Perm", self ctxt at)),
-                                         !!(MacroCall("$Inv", self ctxt at)))))) ])));
-            !!(InhaleS (!!(AndE(!!(MacroCall("$Perm", self ctxt at)),
-                                !!(MacroCall("$Inv", self ctxt at))))));
-          ])
+     !!([],
+        [ !!(FieldAssignS(
+            (self ctxt Source.no_region, !!id),
+            (!!(AddE(!!(FldAcc (self ctxt (s.at), !!id)),
+                     intLitE Source.no_region 1)))));
+          !@(ExhaleS (!@(AndE(!@(MacroCall("$Perm", self ctxt at)),
+                              !@(MacroCall("$Inv", self ctxt at))))));
+          !@(SeqnS (
+              !@([],
+                 [
+                   !@(InhaleS (!@(AndE(!@(MacroCall("$Perm", self ctxt at)),
+                                  !@(AndE(!@(MacroCall("$Inv", self ctxt at)),
+                                          !@(GtCmpE(!@(FldAcc (self ctxt at, !@id)),
+                                               intLitE Source.no_region 0))))))));
+                   !@(FieldAssignS(
+                          (self ctxt at, !@id),
+                          (!@(SubE(!@(FldAcc (self ctxt at, !@id)),
+                                   intLitE at 1)))));
+                   !!! (e.at) (SeqnS stmts);
```

Comment:
Is this maybe confusing things and should really be !@

Reply:
Not so sure.